---
title: "RNA-seq data analysis for grapevine tutorial"
author: "Hongrui Wang"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# RNA-seq data analysis for grapevine

This is an RNA-seq analysis tutorial aiming to provide the entire analysis pipeline and the biological reasoning of all the steps included. The pipeline includes PCA-based outlier filtering, PCA-based general transcriptome characterization, DeSeq2-based normalization, WGCNA-based co-expression network analysis, DeSeq2-based differential expression analysis and GSEA-based pathway enrichment analysis.

The tutorial is based on the method in Wang, H. et al. (2022) ‘Transcriptomic analysis of grapevine in response to ABA application reveals its diverse regulations during cold acclimation and deacclimation’, Fruit Research, 2(1), pp. 1–12. Available at: [here](https://doi.org/10.48130/FruRes-2022-0001)

The demo data used in this tutorial is available at [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE184114)

```{r needed libraries, echo = FALSE, warning=FALSE}
library(tidyverse)
library(ggplot2)
library(DESeq2)
library(ggpubr)
library(tidyr)
library(dplyr)
library(WGCNA)
library(wesanderson)
library(Rmisc)
library(factoextra)
library(scales)
library(Rmisc)
library(tidyr)
library(pracma)
library(data.table)
```

## Input files
sample_metadata.txt is a metadata file providing the treatment of each sample (library) in the experiment. Here, we have two treatments: control and ABA (a plant hormone that tends to paly a role during winter dormancy). In this experiment, we are testing if the application of ABA on woolly grapevine buds would delay budbreak and the transcriptome mechanism associated with the delay. In the metadata file, 'Treatment' indicates the treatment during the experiment, and 'Time' indicates number of hours that the samples has been in treatment
```{r input files, echo = TRUE}
#cold data is the metadata file
coldata <- fread("sample_metadata.txt")
head(coldata)
```
gene_count.txt is a gene count matrix with row.names as gene names and each column as a library.
```{r input files, echo = TRUE}
#cts data is the gene count matrix file
cts_original <- fread("gene_count.txt")
#remove duplications
cts_original = cts_original[!duplicated(cts_original[,1]),]
cts <- cts_original[,-c(1,36)]
row.names(cts) = cts_original$V1
colnames(cts) <- coldata$Sample_name
head(cts, n=50)
```

## Generate normalized gene count to do PCA to identify outliers

```{r outlier removal, echo=FALSE}
coldata$Time <-as.numeric(coldata$Time)
coldata$Treatment <- factor(coldata$Treatment, levels = c("ABA","Control"))

#Deseq2 model: design = ~ Treatment + Time)
dds <- DESeqDataSetFromMatrix(countData = cts, colData = coldata, design = ~ Treatment + Time)

#Low count genefiltering. Criterium: Averagely, at least one count per sample
keep <- rowSums(counts(dds)) >= ncol(cts)
dds <- dds[keep,]


```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
